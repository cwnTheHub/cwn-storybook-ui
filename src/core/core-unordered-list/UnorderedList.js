import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import UnorderedItem from "./UnorderedItem/UnorderedItem";
import { safeRest } from "../../util-helpers";
import { componentWithName } from "../../util-prop-types";
import { list } from "../../shared-styles";
import Box from "../core-box/Box";

const StyledListContainer = styled(Box)({
  ...list.nestedListSpacing,
  "&": {
    paddingLeft: "calc(2rem + 1px)",
  },
});

const injectListStyle = (child, listStyle, size) =>
  React.cloneElement(child, { listStyle, size });

const UnorderedList = ({ listStyle, size, children, ...rest }) => {
  const filteredChildren = React.Children.toArray(children).filter(
    (child) => child
  );
  return (
    <StyledListContainer {...safeRest(rest)} tag="ul" between={2}>
      {React.Children.map(filteredChildren, (child) =>
        injectListStyle(child, listStyle, size)
      )}
    </StyledListContainer>
  );
};

UnorderedList.propTypes = {
  listStyle: PropTypes.oneOf(["circle", "checkmark", "x"]),
  size: PropTypes.oneOf(["small", "medium", "large"]),
  children: componentWithName("UnorderedItem").isRequired,
};

UnorderedList.defaultProps = {
  listStyle: "circle",
  size: "medium",
};

UnorderedList.Item = UnorderedItem;

export default UnorderedList;
