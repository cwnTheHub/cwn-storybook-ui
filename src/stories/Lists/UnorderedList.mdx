import {
  Canvas,
  Meta,
  Title,
  Subtitle,
  Description,
  Primary,
  Controls,
  Stories,
  Story,
} from "@storybook/blocks";

import * as UnorderedListStories from "./UnorderedList.stories";

<Meta of={UnorderedListStories} />
<Title />
`` import {UnorderedList} from '@nds_core/nds' ``

### PROPS & METHODS

<Controls />

### Minimal usage

_Note: An `UnorderedList` should not be used adjacent to an `OrderedList`._

<Canvas of={UnorderedListStories.MinimalUsage} />

### Check list

<Canvas of={UnorderedListStories.WithCheckmark} />

### Error list

<Canvas of={UnorderedListStories.WithError} />
### Mixed List

If you do not want each list item to have the same icon, you may give each item its own icon via the `itemStyle` prop. This will override a `listStyle` set by the parent.

<Canvas of={UnorderedListStories.MixedList} />