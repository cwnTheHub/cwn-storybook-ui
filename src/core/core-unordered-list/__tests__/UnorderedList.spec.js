import React from "react";
import { shallow, mount, render } from "enzyme";
import UnorderedList from "../UnorderedList";

describe("<UnorderedList />", () => {
  const doRender = (overrides = {}) =>
    render(
      <UnorderedList {...overrides}>
        <UnorderedList.Item>Lorem ipsum</UnorderedList.Item>
        {undefined && <UnorderedList.Item>Lorem ipsum</UnorderedList.Item>}
        <UnorderedList.Item>Dolor sit amet</UnorderedList.Item>
      </UnorderedList>
    );

  const doShallow = (overrides = {}) =>
    shallow(
      <UnorderedList {...overrides}>
        <UnorderedList.Item>Lorem ipsum</UnorderedList.Item>
        {undefined && <UnorderedList.Item>Lorem ipsum</UnorderedList.Item>}
        <UnorderedList.Item>Dolor sit amet</UnorderedList.Item>
      </UnorderedList>
    );

  const doMount = (overrides = {}) =>
    mount(
      <UnorderedList {...overrides}>
        <UnorderedList.Item>Lorem ipsum</UnorderedList.Item>
        {undefined && <UnorderedList.Item>Lorem ipsum</UnorderedList.Item>}
        <UnorderedList.Item>Dolor sit amet</UnorderedList.Item>
      </UnorderedList>
    );

  it("renders", () => {
    const unorderedList = doRender();

    expect(unorderedList).toMatchSnapshot();
  });

  it("renders an HTML ul tag", () => {
    const unorderedList = doMount();
    expect(unorderedList.find("ul")).toBeTruthy();
  });

  it("UnorderedList.Item renders an HTML li tag", () => {
    const unorderedListItem = mount(
      <UnorderedList.Item>Some content</UnorderedList.Item>
    );

    expect(unorderedListItem.find("li")).toBeTruthy();
  });

  it("can have a list style", () => {
    let unorderedList = doMount({ listStyle: undefined });
    expect(unorderedList).toMatchSnapshot();

    unorderedList = doMount({ listStyle: "x" });
    expect(unorderedList).toMatchSnapshot();
  });

  it("passes additional attributes to ul element", () => {
    const unorderedList = doShallow({
      id: "the-list",
      "data-some-thing": "some value",
    });

    expect(unorderedList.props().id).toEqual("the-list");
    expect(unorderedList.props()["data-some-thing"]).toEqual("some value");
  });

  it("does not allow custom CSS", () => {
    const unorderedList = doShallow({
      className: "my-custom-class",
      style: { color: "hotpink" },
    });

    expect(unorderedList.props()["className"]).not.toEqual("my-custom-class");
    expect(unorderedList.props()["style"]).toBeFalsy();
  });

  it("renders a mixed list of items", () => {
    const unorderedList = shallow(
      <UnorderedList>
        <UnorderedList.Item itemStyle="checkmark">
          Lorem ipsum
        </UnorderedList.Item>
        <UnorderedList.Item itemStyle="x">Dolor sit amet</UnorderedList.Item>
      </UnorderedList>
    );

    expect(unorderedList).toMatchSnapshot();
  });
});
