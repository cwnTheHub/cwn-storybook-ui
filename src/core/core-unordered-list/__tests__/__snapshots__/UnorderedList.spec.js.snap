// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UnorderedList /> can have a list style 1`] = `
<UnorderedList
  listStyle="circle"
  size="medium"
>
  <Styled(Box)
    between={2}
    tag="ul"
  >
    <Box
      between={2}
      className="sc-eqUAAy eUciOt"
      inline={false}
      tag="ul"
    >
      <styled.div
        between={2}
        className="sc-eqUAAy eUciOt"
        inline={false}
        tag="ul"
      >
        <ul
          className="sc-gEvEer MkeJt sc-eqUAAy eUciOt sc-eqUAAy eUciOt"
        >
          <UnorderedList.Item
            key=".$.0"
            listStyle="circle"
            size="medium"
          >
            <styled.li
              iconStyle="circle"
              size="medium"
            >
              <li
                className="sc-aXZVg chrzLG"
                size="medium"
              >
                Lorem ipsum
              </li>
            </styled.li>
          </UnorderedList.Item>
          <UnorderedList.Item
            key=".$.2"
            listStyle="circle"
            size="medium"
          >
            <styled.li
              iconStyle="circle"
              size="medium"
            >
              <li
                className="sc-aXZVg chrzLG"
                size="medium"
              >
                Dolor sit amet
              </li>
            </styled.li>
          </UnorderedList.Item>
        </ul>
      </styled.div>
    </Box>
  </Styled(Box)>
</UnorderedList>
`;

exports[`<UnorderedList /> can have a list style 2`] = `
<UnorderedList
  listStyle="x"
  size="medium"
>
  <Styled(Box)
    between={2}
    tag="ul"
  >
    <Box
      between={2}
      className="sc-eqUAAy eUciOt"
      inline={false}
      tag="ul"
    >
      <styled.div
        between={2}
        className="sc-eqUAAy eUciOt"
        inline={false}
        tag="ul"
      >
        <ul
          className="sc-gEvEer MkeJt sc-eqUAAy eUciOt sc-eqUAAy eUciOt"
        >
          <UnorderedList.Item
            key=".$.0"
            listStyle="x"
            size="medium"
          >
            <styled.li
              iconStyle="x"
              size="medium"
            >
              <li
                className="sc-aXZVg jvGpWV"
                size="medium"
              >
                Lorem ipsum
              </li>
            </styled.li>
          </UnorderedList.Item>
          <UnorderedList.Item
            key=".$.2"
            listStyle="x"
            size="medium"
          >
            <styled.li
              iconStyle="x"
              size="medium"
            >
              <li
                className="sc-aXZVg jvGpWV"
                size="medium"
              >
                Dolor sit amet
              </li>
            </styled.li>
          </UnorderedList.Item>
        </ul>
      </styled.div>
    </Box>
  </Styled(Box)>
</UnorderedList>
`;

exports[`<UnorderedList /> renders 1`] = `
<ul
  class="sc-gEvEer MkeJt sc-eqUAAy eUciOt sc-eqUAAy eUciOt"
>
  <li
    class="sc-aXZVg chrzLG"
  >
    Lorem ipsum
  </li>
  <li
    class="sc-aXZVg chrzLG"
  >
    Dolor sit amet
  </li>
</ul>
`;

exports[`<UnorderedList /> renders a mixed list of items 1`] = `
<Styled(Box)
  between={2}
  tag="ul"
>
  <UnorderedList.Item
    itemStyle="checkmark"
    key=".$.0"
    listStyle="circle"
    size="medium"
  >
    Lorem ipsum
  </UnorderedList.Item>
  <UnorderedList.Item
    itemStyle="x"
    key=".$.1"
    listStyle="circle"
    size="medium"
  >
    Dolor sit amet
  </UnorderedList.Item>
</Styled(Box)>
`;
